components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/token/AbstractJWTGenerator.java

```diff
@@ -0,0 +1,633 @@
+                    String jksName = ksName + ".jks";
+                    //obtain private key
+                    //TODO: maintain a hash map with tenants' private keys after first initialization
+                    privateKey = tenantKSM.getPrivateKey(jksName, tenantDomain);
+                }else{
+                    try{
+                        privateKey = tenantKSM.getDefaultPrivateKey();
+                    }catch (Exception e){
+                        log.error("Error while obtaining private key for super tenant",e);
+                    }
+                }
+                if (privateKey != null) {
+                    privateKeys.put(tenantId, privateKey);
+                }
+            } else {
+                privateKey = privateKeys.get(tenantId);
+            }
+
+            //initialize signature with private key and algorithm
+            Signature signature = Signature.getInstance(signatureAlgorithm);
+            signature.initSign((PrivateKey) privateKey);
+
+            //update signature with data to be signed
+            byte[] dataInBytes = assertion.getBytes();
+            signature.update(dataInBytes);
+
+            //sign the assertion and return the signature
+            byte[] signedInfo = signature.sign();
+            return signedInfo;
+
+        } catch (NoSuchAlgorithmException e) {
+            String error = "Signature algorithm not found.";
+            //do not log
+            throw new AppManagementException(error);
+        } catch (InvalidKeyException e) {
+            String error = "Invalid private key provided for the signature";
+            //do not log
+            throw new AppManagementException(error);
+        } catch (SignatureException e) {
+            String error = "Error in signature";
+            //do not log
+            throw new AppManagementException(error);
+        } catch (AppManagementException e) {
+            //do not log
+            throw new AppManagementException(e.getMessage());
+        }
+    }
+
+  /**
+     * Helper method to add public certificate to JWT_HEADER to signature verification.
+     *
+     * @param endUserName
+     * @throws org.wso2.carbon.appmgt.api.AppManagementException
+     */
+    private String addCertToHeader(String endUserName) throws AppManagementException {
+
+        try {
+            //get tenant domain
+            String tenantDomain = MultitenantUtils.getTenantDomain(endUserName);
+            //get tenantId
+            int tenantId = getTenantId(endUserName);
+            Certificate publicCert = null;
+
+            if (!(publicCerts.containsKey(tenantId))) {
+                //get tenant's key store manager
+                AppManagerUtil.loadTenantRegistry(tenantId);
+                KeyStoreManager tenantKSM = KeyStoreManager.getInstance(tenantId);
+
+                KeyStore keyStore = null;
+                if(!tenantDomain.equals(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME)){
```

Comment:
Can re-factor so that getting the keyStore can be extracted to single method to be used  in both places.